fe/fe-core/src/main/java/com/starrocks/authentication/AuthenticationManager.java

```diff
@@ -96,16 +117,52 @@ public String getDefaultPlugin() {
         return DEFAULT_PLUGIN;
     }
 
+    /**
+     * If someone login from 10.1.1.1 with name "test_user", the matching UserIdentity can be sorted in the below order
+     * 1. test_user@10.1.1.1
+     * 2. test_user@["hostname"], in which "hostname" can be resolved to 10.1.1.1.
+     * If multiple hostnames matche the login ip, just return one randomly.
+     * 3. test_user@%, as a fallback.
+     */
+    private Integer scoreUserIdentityHost(UserIdentity userIdentity) {
+        // ip(1) > hostname(2) > %(3)
+        if (userIdentity.isDomain()) {
+            return 2;
+        }
+        if (userIdentity.getHost().equals(UserAuthenticationInfo.ANY_HOST)) {
+            return 3;
+        }
+        return 1;
+    }
+
+    private boolean match(String remoteUser, String remoteHost, boolean isDomain, UserAuthenticationInfo info) {
+        // quickly filter unmatched entries by user name
+        if (!info.matchUser(remoteUser)) {
+            return false;
+        }
+        if (isDomain) {
+            // check for resolved ips
+            Set<String> ipSet = hostnameToIpSet.get(info.getOrigHost());
```

Comment:
do we need hostnameToIpLock.readLock() here?

Reply:
My mistake. Done.